gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1228 @@
+								 RelationGetDescr(rel), &isnull);
+		if (!isnull)
+		{
+			ArrayType  *array;
+			Datum	   *elems;
+			int			nelems;
+
+			array = DatumGetArrayTypeP(str_datum);
+			deconstruct_array(array, TEXTOID, -1, false, 'i',
+							  &elems, NULL, &nelems);
+			for (int i = 0; i < nelems; i++)
+			{
+				char	   *str = TextDatumGetCString(elems[i]);
+
+				if (strncmp(str, "arenadata_toolkit.tracking_schemas=", 35) == 0)
+				{
+					current_schemas = pstrdup(str + 35);
+					break;
+				}
+				pfree(str);
+			}
+		}
+	}
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	new_schemas = add_or_remove_schema(current_schemas, schemaName, reg);
+
+	stmt.type = T_AlterDatabaseSetStmt;
+	stmt.dbname = get_database_name(dbid);
+
+	if (stmt.dbname == NULL)
+		elog(ERROR, "[arenadata_toolkit] database %u does not exist", dbid);
+
+	stmt.setstmt = &v_stmt;
+
+	v_stmt.type = T_VariableSetStmt;
+	v_stmt.name = "arenadata_toolkit.tracking_schemas";
+	v_stmt.is_local = false;
+
+	arg.type = T_A_Const;
+	arg.val.type = T_String;
+	arg.val.val.str = new_schemas;
+	arg.location = -1;
+
+	if (new_schemas == NULL)
+	{
+		/*
+		 * If new_schemas is NULL, we're removing the last schema, so let's
+		 * just RESET the variable
+		 */
+		v_stmt.kind = VAR_RESET;
+		v_stmt.args = NIL;
+	}
+	else
+	{
+		v_stmt.kind = VAR_SET_VALUE;
+		v_stmt.args = list_make1(&arg);
+	}
+
+	tf_guc_unlock_schemas_once();
+
+	AlterDatabaseSet(&stmt);
+
+	if (current_schemas)
+		pfree(current_schemas);
+	if (new_schemas)
+		pfree(new_schemas);
+}
+
+Datum
+tracking_register_schema(PG_FUNCTION_ARGS)
+{
+	const char *schema_name = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = PG_GETARG_OID(1);
+
+	dbid = dbid == InvalidOid ? MyDatabaseId : dbid;
+
+	if (!SearchSysCacheExists1(NAMESPACENAME, CStringGetDatum(schema_name)))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema %s does not exist", schema_name)));
+
+	elog(LOG, "[arenadata_toolkit] registering schema %s in database %u for tracking", schema_name, dbid);
+
+	track_schema(schema_name, dbid, true);
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_unregister_schema(PG_FUNCTION_ARGS)
+{
+	const char *schema_name = NameStr(*PG_GETARG_NAME(0));
+	Oid			dbid = PG_GETARG_OID(1);
+
+	dbid = dbid == InvalidOid ? MyDatabaseId : dbid;
+
+	if (!SearchSysCacheExists1(NAMESPACENAME, CStringGetDatum(schema_name)))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("schema with OID %s does not exist", schema_name)));
+
+	elog(LOG, "[arenadata_toolkit] registering schema %s in database %u for tracking", schema_name, dbid);
+
+	track_schema(schema_name, dbid, false);
+
+	PG_RETURN_BOOL(true);
+}
+
+static bool
+is_valid_relkind(char relkind)
+{
+	return (relkind == 'r' || relkind == 'i' || relkind == 'S' ||
+			relkind == 't' || relkind == 'v' || relkind == 'c' ||
+			relkind == 'f' || relkind == 'u' || relkind == 'm' ||
+			relkind == 'o' || relkind == 'b' || relkind == 'M');
```

Comment:
Maybe this would be cleaner and faster by some nanoseconds?
```suggestion
	switch (relkind) {
	case 'r':
	case 'i':
	case 'S':
	case 't':
	case 'v':
	case 'c':
	case 'f':
	case 'u':
	case 'm':
	case 'o':
	case 'b':
	case 'M':
		return true;
	default:
		return false;
	}
```